File: lightbite-rn/src/core/api/refresh.interceptor.ts
import axios, { type AxiosError, type InternalAxiosRequestConfig } from 'axios';

import { setAuthToken } from '@/core/api/auth.interceptor';
import { SecureStorage } from '@/core/storage/secure-storage';

let isRefreshing = false;
let pendingQueue: {
  resolve: (token: string) => void;
  reject: (error: unknown) => void;
}[] = [];

function processQueue(error: unknown, token: string | null): void {
  pendingQueue.forEach(({ resolve, reject }) => {
    if (error) {
      reject(error);
    } else if (token) {
      resolve(token);
    }
  });
  pendingQueue = [];
}

export async function refreshInterceptor(error: AxiosError): Promise<InternalAxiosRequestConfig> {
  const originalRequest = error.config as InternalAxiosRequestConfig & { _retry?: boolean };

  // Only handle 401 errors, and don't retry the refresh endpoint itself
  if (error.response?.status !== 401 || originalRequest.url?.includes('/auth/refresh')) {
    return Promise.reject(error);
  }

  // Don't retry if we already tried
  if (originalRequest._retry) {
    return Promise.reject(error);
  }

  const refreshToken = await SecureStorage.getRefreshToken();
  if (!refreshToken) {
    return Promise.reject(error);
  }

  if (isRefreshing) {
    // Queue this request until refresh completes
    return new Promise<InternalAxiosRequestConfig>((resolve, reject) => {
      pendingQueue.push({
        resolve: (token: string) => {
          originalRequest.headers.Authorization = `Bearer ${token}`;
          resolve(originalRequest);
        },
        reject,
      });
    });
  }

  originalRequest._retry = true;
  isRefreshing = true;

  try {
    const response = await axios.post<{
      data: { access_token: string; refresh_token: string; expires_in: number };
    }>(
      `${originalRequest.baseURL}/auth/refresh`,
      { refresh_token: refreshToken },
      { headers: { 'Content-Type': 'application/json' } },
    );

    const { access_token, refresh_token } = response.data.data;

    await SecureStorage.setAccessToken(access_token);
    await SecureStorage.setRefreshToken(refresh_token);
    setAuthToken(access_token);

    processQueue(null, access_token);
    isRefreshing = false;

    originalRequest.headers.Authorization = `Bearer ${access_token}`;
    return originalRequest;
  } catch (refreshError) {
    processQueue(refreshError, null);
    isRefreshing = false;

    await SecureStorage.clearAll();
    setAuthToken(null);

    return Promise.reject(refreshError);
  }
}

File: lightbite-app/lib/core/network/refresh_interceptor.dart
import 'package:dio/dio.dart';
import 'package:lightbite_app/core/bus/app_events.dart';
import 'package:lightbite_app/core/bus/event_bus.dart';
import '../storage/secure_storage.dart';

/// Dio interceptor that handles token refresh on 401 responses.
///
/// Implements a queue-safe refresh mechanism: when multiple concurrent
/// requests receive 401, only one refresh call is made. Queued requests
/// are retried after the refresh succeeds, or all fail if it fails.
class RefreshInterceptor extends Interceptor {
  RefreshInterceptor(this._storage, this._dio);

  final SecureStorage _storage;
  final Dio _dio;

  /// Prevents concurrent refresh attempts.
  bool _isRefreshing = false;

  /// Requests that arrived while a refresh was in progress.
  final List<({DioException error, ErrorInterceptorHandler handler})>
      _pendingQueue = [];

  @override
  void onError(DioException err, ErrorInterceptorHandler handler) async {
    if (err.response?.statusCode != 401) {
      return handler.next(err);
    }

    // Don't retry the refresh endpoint itself — avoids infinite loop
    if (err.requestOptions.path == '/auth/refresh') {
      await _storage.clearAll();
      return handler.next(err);
    }

    // Queue this request if a refresh is already in progress
    if (_isRefreshing) {
      _pendingQueue.add((error: err, handler: handler));
      return;
    }

    _isRefreshing = true;
    try {
      final refreshToken = await _storage.getRefreshToken();
      if (refreshToken == null) {
        await _storage.clearAll();
        appBus.fire(const SessionExpiredEvent());
        return handler.next(err);
      }

      final response = await Dio(
        BaseOptions(baseUrl: _dio.options.baseUrl),
      ).post('/auth/refresh', data: {'refresh_token': refreshToken});

      final newToken = response.data['data']['access_token'] as String;
      final newRefresh = response.data['data']['refresh_token'] as String?;

      await _storage.saveToken(newToken);
      if (newRefresh != null) {
        await _storage.saveRefreshToken(newRefresh);
      }

      // Retry the original request with new token
      final retryOptions = err.requestOptions;
      retryOptions.headers['Authorization'] = 'Bearer $newToken';
      final retryResponse = await _dio.fetch(retryOptions);
      handler.resolve(retryResponse);

      // Retry all queued requests
      for (final entry in _pendingQueue) {
        final retryOpts = entry.error.requestOptions;
        retryOpts.headers['Authorization'] = 'Bearer $newToken';
        final queuedRetry = await _dio.fetch(retryOpts);
        entry.handler.resolve(queuedRetry);
      }
    } on DioException catch (refreshErr) {
      // Only clear session if the refresh endpoint itself returned 401.
      // Network errors should NOT log the user out.
      if (refreshErr.response?.statusCode == 401) {
        await _storage.clearAll();
        appBus.fire(const SessionExpiredEvent());
      }
      // Fail the original request
      handler.next(err);
      // Fail all queued requests
      for (final entry in _pendingQueue) {
        entry.handler.next(entry.error);
      }
    } catch (_) {
      // Unexpected error (parse, runtime) — don't log out, just fail.
      handler.next(err);
      for (final entry in _pendingQueue) {
        entry.handler.next(entry.error);
      }
    } finally {
      _isRefreshing = false;
      _pendingQueue.clear();
    }
  }
}
